File: elixir/twice_linear.ex
defmodule Twice do

    def dbl_linear(n) do
     y = fn x -> 2*x+1 end
     z = fn x -> 3*x+1 end
     Enum.reduce(1..n, {[1], 0, 0}, fn _, acc ->
       yaux = y.(Enum.at(elem(acc, 0), elem(acc, 1)))
       zaux = z.(Enum.at(elem(acc, 0), elem(acc, 2)))
       cond do
        yaux <= zaux ->
         acc = {elem(acc, 0) ++ [yaux], elem(acc, 1) + 1, elem(acc, 2)}
         if yaux == zaux do
         {elem(acc, 0), elem(acc, 1), elem(acc, 2) + 1}
        else
         acc
        end
        true ->
         {elem(acc, 0) ++ [zaux], elem(acc, 1), elem(acc, 2) + 1}
       end
     end)
     |> elem(0)
     |> List.last()
     end
end

File: kotlin/twice_linear.kt
package dbllinear
fun dblLinear(n:Int):Int {
    val u : MutableList<Int> = mutableListOf(1)
    var x:Int = 0;
    var y:Int = 0;
    val fx = {j:Int -> 2*j + 1}
    val fy = {j:Int -> 3*j + 1}

    for (i in 1..n){
        var valx = fx(u[x])
        var valy = fy(u[y])
        if (valx <= valy){
            u.add(valx)
            x++
            if (valx == valy){
                y++
            }
        } else {
            u.add(valy)
            y++
        }
    }
    return u.last()
}
